odl/solvers/functional/functional.py

```diff
@@ -0,0 +1,896 @@
+        return self.left_func.proximal
+
+    @property
+    def convex_conj(self):
+        """Convex conjugate functional of FunctionalScalarSum."""
+        return self.left.convex_conj - self.scalar
+
+
+class FunctionalTranslation(Functional):
+
+    """Implementation of the translated functional.
+
+    Given a functional ``f`` and an element ``translation`` in the domain of
+    ``f``, this corresponds to the functional ``f(. - translation)``.
+    """
+
+    def __init__(self, func, translation):
+        """Initialize a new instance.
+
+        Given a functional ``f(.)`` and a vector ``translation`` in the domain
+        of ``f``, this corresponds to the functional ``f(. - translation)``.
+
+        Parameters
+        ----------
+        func : `Functional`
+            Functional which is to be translated.
+        translation : `domain` element
+            The translation.
+        """
+        if not isinstance(func, Functional):
+            raise TypeError('`func` {!r} not a `Functional` instance'
+                            ''.format(func))
+
+        translation = func.domain.element(translation)
+
+        super().__init__(space=func.domain, linear=False,
+                         grad_lipschitz=func.grad_lipschitz)
+
+        # TODO: Add case if we have translation -> scaling -> translation?
+        if isinstance(func, FunctionalTranslation):
+            self.__original_func = func.original_func
+            self.__translation = func.translation + translation
+
+        else:
+            self.__original_func = func
+            self.__translation = translation
+
+    @property
+    def original_func(self):
+        """The original functional that has been translated."""
+        return self.__original_func
+
+    @property
+    def translation(self):
+        """The translation."""
+        return self.__translation
+
+    def _call(self, x):
+        """Evaluate the functional in a point ``x``."""
+        return self.original_func(x - self.translation)
+
+    @property
+    def gradient(self):
+        """Gradient operator of the functional."""
+        return (self.original_func.gradient *
+                (IdentityOperator(self.domain) - self.translation))
+
+    @property
+    def proximal(self):
+        """Proximal factory of the translated functional.
+
+        See Also
+        --------
+        proximal_translation
+        """
+        return proximal_translation(self.original_func.proximal,
+                                    self.translation)
+
+    @property
+    def convex_conj(self):
+        """Convex conjugate functional of the translated functional."""
+        return FunctionalLinearPerturb(
+            self.original_func.convex_conj,
+            self.translation)
+
+
+class FunctionalLinearPerturb(Functional):
+
+    """The ``Functional`` representing ``f(.) + <linear_term, .>``.
+
+    Notes
+    -----
+    Given a functional :math:`f`, the convex conjugate of a translated version
```

Comment:
This section feels out of place, perhaps move it to `FunctionalTranslation.convex_conj`?